model/src/main/java/org/activityinfo/model/formTree/LookupKey.java

```diff
@@ -71,27 +67,54 @@
      */
     String keyLabel;
 
+    /**
+     * A key cannot know itself whether it is leaf key - keys both with and without children can be valid candidates
+     * for a leaf key depending on what level we are selecting from. This is set by {@link LookupKeySet} if it has
+     * determined the key is a leaf key.
+     */
+    boolean leaf = false;
+
+    /**
+     * Defines a parent of a key, including the field path for navigating to the parent in the form tree
+     */
+    protected static class ParentKey {
```

Comment:
Why `protected` and not `private`? I don't think LookupKey is a class that we would want to leave open to extension. (You could even mark it `final`)

Reply:
I left it as protected as it needs to be accessed by the LookupKeySet while it constructs each LookupKey - see [here](https://github.com/bedatadriven/activityinfo/blob/65d04497275f45fd9e8de8e91354177665905f47/model/src/main/java/org/activityinfo/model/formTree/LookupKeySet.java#L111). 

I can alter this to `protected final` so that it is available to LookupKeySet but can't be extended.

Reply:
Ah, I actually didn't realize that `protected` members were accessible by other classes in the same package. You can also omit the access modifier, which result in the method having "package-private" access, a status slightly more restrictive than "protected" as it does not include subclasses:
https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html

Reply:
OK I'll leave it as `final` then but make it `package-private`